### PR TITLE
Replace NoStackTraceThrowable with NoStackTraceException #4854

### PR DESCRIPTION
**Pull request by magicprinc**
See https://github.com/eclipse-vertx/vert.x/issues/4854

### CODE REVIEW

**Comment by magicprinc** [comment]
NoStackTraceThrowable is marked as @Deprecated with javadoc ```@deprecated Use {@link NoStackTraceException} instead.```, but not deleted.
It allows fixing other vert.x modules eventually without breaking or fixing all at once.

File: src/main/java/io/vertx/core/impl/NoStackTraceThrowable.java [hunk 1]
```diff
@@ -12,11 +12,12 @@
...
 
 /**
+ * @deprecated Use {@link NoStackTraceException} instead.
  * @author <a href="http://tfox.org">Tim Fox</a>
  */
-public class NoStackTraceThrowable extends Throwable {
-
-  public NoStackTraceThrowable(String message) {
-    super(message, null, false, false);
+@Deprecated
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/NoStackTraceThrowable.java:18]: why do we need to deprecate that ?

Comment (magicprinc) [reply src/main/java/io/vertx/core/impl/NoStackTraceThrowable.java:18]: Do you mean:
1) delete it
or
2) keep it without detraction?
-
1) Delete: Great Idea! I completely agree! But this class is used in two other modules:
https://github.com/search?q=org%3Aeclipse-vertx+NoStackTraceThrowable&type=code
(vertx-junit5 and vertx-sql-client)
And I don't have time to fix them too 🙏
With the deprecation, you can defer this fix.

2) Keep it
But why?
a) It duplicates the fuctionality
b) Direct Throwable subclass is bad. Evenn here in this project there were already problems with it
https://github.com/eclipse-vertx/vert.x/issues/4082

Comment (magicprinc) [reply src/main/java/io/vertx/core/impl/NoStackTraceThrowable.java:18]: Extending Throwable directly is generally considered a bad practice in Java for several reasons:

1) **Hierarchy of Exceptions**: In Java, Throwable is the superclass of all errors and exceptions. It has two direct subclasses: Error and Exception [baeldung.com](https://www.baeldung.com/java-catch-throwable-bad-practice). Error and its subclasses represent serious problems that a reasonable application should not try to catch because they indicate conditions that an application usually cannot or should not recover from, such as OutOfMemoryError or StackOverflowError. On the other hand, Exception and its subclasses represent conditions that a reasonable application might want to catch and handle, such as FileNotFoundException or NullPointerException [baeldung.com](https://www.baeldung.com/java-catch-throwable-bad-practice). By extending Throwable directly, you are bypassing this hierarchy and potentially creating confusion about the nature and severity of the condition that your class represents.

2) **Exception Handling**: Exceptions in Java are meant to be caught and handled appropriately. The general rule in handling exceptions is that the try-catch block must be as specific as possible in catching exceptions. That is, a catch-all scenario must be avoided [baeldung.com](https://www.baeldung.com/java-catch-throwable-bad-practice). By extending Throwable, you might be creating a type that can be caught by a general catch (Throwable t) block, leading to inappropriate handling of different types of errors and exceptions.

3) **Code Clarity**: Many applications assume only Exception and Error exist as subclasses of Throwable [devtut.github.io](https://devtut.github.io/java/java-pitfalls-exception-usage.html). Creating a new class that extends Throwable directly can be confusing for other developers who might expect your class to be a subclass of Exception or Error. It's more clear and conventional to extend Exception or one of its subclasses when you need to create a custom checked exception, and to extend RuntimeException or one of its subclasses when you need to create a custom unchecked exception.

4) Checked vs Unchecked: Classes directly extending Throwable are checked, as Exception is, so the compiler will enforce that this Throwable is declared to be thrown or caught. This can lead to verbose and cluttered code if the Throwable is used frequently [blog.vanillajava.blog](https://blog.vanillajava.blog/2021/12/unusual-java-stacktrace-extends.html).

**In conclusion**, while Java allows you to extend Throwable directly, it's generally better to extend Exception, RuntimeException, or one of their subclasses when you need to create a custom exception. This approach respects the established hierarchy of exceptions in Java, leads to clearer and more maintainable code, and allows for more appropriate and specific exception handling.

Comment (vietj) [reply src/main/java/io/vertx/core/impl/NoStackTraceThrowable.java:18]: I mean that removing it or changing its inheritance can be a un whishable breaking change for user applications.

Comment (magicprinc) [reply src/main/java/io/vertx/core/impl/NoStackTraceThrowable.java:18]: I have returned the inheritance hierarchy: we will not break anything. All 

Annotation ```@Deprecated``` (with detailed javadoc) doesn't harm any project using this class, but clearly shows a better solution.

Comment (magicprinc) [reply src/main/java/io/vertx/core/impl/NoStackTraceThrowable.java:20]: I have finally found the best solution:
```VertxException``` → ```NoStackTraceThrowable``` → ```NoStackTraceException```

If some code catches NoStackTraceThrowable, it won't break (NoStackTraceException is NoStackTraceThrowable), but shows deprecated waring.
After all code migrates to NoStackTraceException, NoStackTraceThrowable can be removed:
```VertxException``` → ```NoStackTraceException```.
Nobody will miss changing parent of NoStackTraceThrowable from Throwable to RuntimeException (VertxException).

I can't imagine situation, where it can break code

- catch (NoStackTraceThrowable e) - works
- catch (NoStackTraceThrowable e) - works
- catch (Throwable e) - works

Comment (vietj) [reply src/main/java/io/vertx/core/impl/NoStackTraceThrowable.java:20]: VertxException will not extend NoStackTraceThrowable as seducing as it can look

Comment (magicprinc) [reply src/main/java/io/vertx/core/impl/NoStackTraceThrowable.java:20]: The point is:
- extending Throwable is bad practice.
- there were and will be problems with such practice.
- this old exception duplicates functionality of a new one (NoStackTraceThrowable ≈ NoStackTraceException)

I hope you know the better solution. this decision is entirely up to you 🙏👍

Comment (vietj) [reply src/main/java/io/vertx/core/impl/NoStackTraceThrowable.java:20]: ok let's deprecate NoStrackTraceThrowable and redirect to use NoStackTraceException instead for now

Comment (vietj) [reply src/main/java/io/vertx/core/impl/NoStackTraceThrowable.java:20]: I think we should not deprecate NoStackTraceThrowable itself, but instead its constructors

Comment (magicprinc) [reply src/main/java/io/vertx/core/impl/NoStackTraceThrowable.java]: A great idea! Done (PR is updated)
